### PR TITLE
code suggestions for non-shorthand field pattern, no-mangle lints

### PR DESCRIPTION
**Pull request by zackmdavis**
continuing in the spirit of #44942

moar_lint_suggestions

r? @estebank

### CODE REVIEW

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1130,35 +1133,55 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidNoMangleItems {
...
                 if attr::contains_name(&it.attrs, "no_mangle") &&
                    !cx.access_levels.is_reachable(it.id) {
-                    let msg = format!("static {} is marked #[no_mangle], but not exported",
-                                      it.name);
-                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, &msg);
+                       let msg = format!("static {} is marked #[no_mangle], but not exported",
+                                         it.name);
+                       let mut err = cx.struct_span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, &msg);
+                       let insertion_span = it.span.with_hi(BytePos(it.span.lo().0));
+                       err.span_suggestion(insertion_span, "try making public", "pub ".to_owned());
```
Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: nit: "try making it public"

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -1130,35 +1133,55 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidNoMangleItems {
...
         match it.node {
             hir::ItemFn(.., ref generics, _) => {
-                if attr::contains_name(&it.attrs, "no_mangle") &&
-                   !attr::contains_name(&it.attrs, "linkage") {
+                if let Some(no_mangle_attr) = attr::find_by_name(&it.attrs, "no_mangle") {
+                    if attr::contains_name(&it.attrs, "linkage") {
+                        return;
+                    }
                     if !cx.access_levels.is_reachable(it.id) {
                         let msg = format!("function {} is marked #[no_mangle], but not exported",
```
Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: drive by fix: can you remove the function names? Most diagnostics pointing to a definition don't mention the name again.

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -1130,35 +1133,55 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidNoMangleItems {
...
                     }
                 }
             }
             hir::ItemStatic(..) => {
                 if attr::contains_name(&it.attrs, "no_mangle") &&
                    !cx.access_levels.is_reachable(it.id) {
-                    let msg = format!("static {} is marked #[no_mangle], but not exported",
-                                      it.name);
-                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, &msg);
+                       let msg = format!("static {} is marked #[no_mangle], but not exported",
```
Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: same here (wrt name repetition in diagnostics)

File: src/librustc_lint/builtin.rs [hunk 4]
```diff
@@ -1130,35 +1133,55 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidNoMangleItems {
...
+                                         it.name);
+                       let mut err = cx.struct_span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, &msg);
+                       let insertion_span = it.span.with_hi(BytePos(it.span.lo().0));
```
Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: this should be `it.span.with_hi(it.span.lo())`

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by zackmdavis** [comment]
(force-push addresses @oli-obk's comments)

**Comment by zackmdavis** [comment]
(and the ensuing compile-fail failures reported by our mutual friend Travis)

**Comment by estebank** [comment]
R=me once the merge conflict is fixed

**Comment by zackmdavis** [comment]
(and @bors's treachery)

**Comment by oli-obk** [comment]
Unrelated and spurious?

```
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/coff2yaml.cpp.o
[00:17:26] clang: error: unable to execute command: Segmentation fault: 11
[00:17:26] clang: error: linker command failed due to signal (use -v to see invocation)
```

Segfault looks bad though

**Comment by kennytm** [comment]
@bors retry #38878

<details>

```
[00:17:25] Scanning dependencies of target obj2yaml
[00:17:25] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/obj2yaml.cpp.o
[00:17:26] [ 98%] Linking CXX executable ../../bin/llvm-symbolizer
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/coff2yaml.cpp.o
[00:17:26] clang: error: unable to execute command: Segmentation fault: 11
[00:17:26] clang: error: linker command failed due to signal (use -v to see invocation)
[00:17:26] make[3]: *** [bin/llvm-symbolizer] Error 254
[00:17:26] make[2]: *** [tools/llvm-symbolizer/CMakeFiles/llvm-symbolizer.dir/all] Error 2
[00:17:26] make[2]: *** Waiting for unfinished jobs....
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/dwarf2yaml.cpp.o
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/elf2yaml.cpp.o
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/macho2yaml.cpp.o
[00:17:26] [ 98%] Building CXX object tools/obj2yaml/CMakeFiles/obj2yaml.dir/Error.cpp.o
[00:17:27] [ 98%] Linking CXX executable ../../bin/obj2yaml
[00:17:27] [ 98%] Built target obj2yaml
[00:17:27] make[1]: *** [all] Error 2
[00:17:27] thread 'main' panicked at '
[00:17:27] command did not execute successfully, got: exit code: 2
[00:17:27] 
[00:17:27] build script failed, must exit now', /Users/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.26/src/lib.rs:599:4
[00:17:27] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

</details>

**Comment by zackmdavis** [comment]
goddamit @bors I _really_ wanted that [T-shirt](https://hacktoberfest.digitalocean.com/)

**Comment by kennytm** [comment]
@bors retry 

* The macs 6.5-hour timed out. Still Travis's issue.

**Comment by estebank** [comment]
It looks to me that travis/mac will timeout again :-/

**Pull request merged**